pkg/registry/registry_docker.go

```diff
@@ -0,0 +1,151 @@
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"text/template"
+
+	"github.com/docker/docker/api/types"
+	"github.com/giantswarm/microerror"
+
+	"github.com/giantswarm/retagger/pkg/images"
+)
+
+func (r *Registry) Login() error {
+	r.logger.Log("level", "debug", "message", fmt.Sprintf("logging in to %s registry", r.host))
+
+	authConfig := types.AuthConfig{
+		Username:      r.username,
+		Password:      r.password,
+		ServerAddress: r.host,
+	}
+	res, err := r.docker.RegistryLogin(context.Background(), authConfig)
+	if err != nil {
+		return microerror.Maskf(err, "could not login to registry")
+	} else if res.Status != "Login Succeeded" {
+		return microerror.Mask(dockerError)
+	}
+
+	r.registryAuth = res.IdentityToken
+
+	return nil
+}
+
+func (r *Registry) PullImage(image string, sha string) error {
+	if sha == "" {
+		return microerror.Maskf(invalidArgumentError, "%s SHA should not be empty", image)
+	}
+
+	shaName := images.ShaName(image, sha)
+
+	r.logger.Log("level", "debug", "message", fmt.Sprintf("executing: docker pull %s", shaName))
+
+	res, err := r.docker.ImagePull(context.Background(), shaName, types.ImagePullOptions{})
+	if err != nil {
+		return microerror.Maskf(err, "could not pull image")
+	}
+	defer res.Close()
+	err = r.logDocker(res)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Registry) TagSha(sourceImage, sha, destinationImage, destinationTag string) (string, error) {
+	imageSha := images.ShaName(sourceImage, sha)
+	retaggedNameWithTag := fmt.Sprintf("%s:%s", destinationImage, destinationTag)
+
+	r.logger.Log("level", "debug", "message", fmt.Sprintf("executing: docker tag %s %s", imageSha, retaggedNameWithTag))
+
+	err := r.docker.ImageTag(context.Background(), imageSha, retaggedNameWithTag)
+	if err != nil {
+		return "", microerror.Maskf(err, "could not tag image")
+	}
+
+	return retaggedNameWithTag, nil
+}
+
+func (r *Registry) PushImage(destinationImage, destinationTag string) error {
+	imageTag := fmt.Sprintf("%s:%s", destinationImage, destinationTag)
+
+	r.logger.Log("level", "debug", "message", fmt.Sprintf("executing: docker push %s", imageTag))
+
+	opts := types.ImagePushOptions{
+		All:          true,
+		RegistryAuth: r.getRegistryAuthBase64(),
+	}
+	res, err := r.docker.ImagePush(context.Background(), imageTag, opts)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	defer res.Close()
+	err = r.logDocker(res)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Registry) RebuildImage(sourceImage, sha, destinationImage, destinationTag string, dockerfileOptions []string) (string, error) {
+	retaggedNameWithTag := fmt.Sprintf("%s:%s", destinationImage, destinationTag)
+
+	dockerfile := Dockerfile{
+		BaseImage:         sourceImage,
+		Sha:               sha,
+		DockerfileOptions: dockerfileOptions,
+		Tag:               destinationTag,
+	}
+
+	f, err := os.Create(fmt.Sprintf("Dockerfile-%s", destinationTag))
```

Comment:
Not hyper important but having a `key.go` with stuff like `fmt.Sprintf("Dockerfile-%s", destinationTag)` as its own little function is pretty handy. 
We do this in operators as these functions can then be reused (see line 120)